autowarefoundation/autoware.universe
# feat(bluetooth_monitor): add functionality to monitor Bluetooth connection

## Description

Add a feature to monitor a Bluetooth connection to a wireless controller.

## Related links

https://github.com/autowarefoundation/autoware.universe/issues/859

## Tests performed

0. Assign capability to `l2ping_service`.
   ```
   sudo setcap 'cap_net_raw+eip' ./build/bluetooth_monitor/l2ping_service
   ```
1. Connect a Bluetooth device.

2. Run `l2ping_service` and `bluetooth_monitor`.
   ```
   ./build/bluetooth_monitor/l2ping_service
   ros2 launch bluetooth_monitor bluetooth_monitor.launch.xml
   ```
3. Run RQt Runtime Monitor.
   ```
   ros2 run rqt_runtime_monitor rqt_runtime_monitor 
   ```
4. Confirm the diag status is `OK`. 
![image](https://user-images.githubusercontent.com/57388357/167076554-1d89ab8c-06e3-4e29-b029-3eb7c76fdf78.png)

5. Keep a Bluetooth device away, then confirm the diag status is `ERROR`. 
![image](https://user-images.githubusercontent.com/57388357/167076705-53d7017d-32f7-45cd-9e68-a164606d835c.png)

6. Bring the device closer, then confirm the diag status returns to `OK`. 
![image](https://user-images.githubusercontent.com/57388357/167076759-0f774aaf-3cb6-404f-8986-5f0d05b9d0fa.png)

## Notes for reviewers

n/a

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.


## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.


## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.


After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/
